### ISSUE TITLE
Use VS Code Remote with offline client

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode-remote-release#1242]
https://github.com/microsoft/vscode-remote-release/issues/15 improved the experience when connecting to a host that isn't connected to the internet. Some users also need to connect from a client that is not connected to the internet. This would require a way for them to manually download the server and set it up on the host, or give it to the client to transfer to each host. And the user would have to do this after every vscode update.

**Comment by WSLUser** [comment]
1.) As mentioned in #15, settings should be provided to choose the server to download (in the Settings UI as well as obviously `settings.json`). Whenever VSCode updates, the server would be updated as well. 

2.) Users should probably be able to choose to download more than one server. Then when client connects to the host, the extension does a check to see which server it should use. Discovering this, it then transfers the server over to the host and does the install. 

3.) Extensions should be downloaded for use by the server at the same time as the server downloads based once again upon the architecture.

For now, we just need the 1 done followed by 3. 2 is an eventual need for some users but probably not as much of a need as 1 and 3.

**Comment by ttutko** [comment]
I mentioned this in #15 but I would like to mention it here too. The C++ extension from Microsoft seems to handle this well by providing the offline installers for the components that get downloaded (https://github.com/microsoft/vscode-cpptools). 

Granted remote plugin is slightly trickier in that it really needs to exist on two machines but I would think that we could use a similar mechanism to download whichever permutations we want/need to our local machines (or a network share) and then use a setting to specify the location from which to pull the server files. The setting could default to empty or to some other value to indicate that it should be downloaded from github as it does now but if it's set to a path (local or network share), then it should look for the needed server in that location. I think it would be reasonable for the plugin to do a quick (short timeout) check for internet to see if there are any newer versions of the server and produce a warning message that newer versions are available but should still use the version of the server from whatever location is specified.

**Comment by strike5150** [comment]
Why wouldn't we just cache whatever we determine at runtime after an update and use the cached copy until the next update.  Most users will have connectivity at some point, its really just about getting this working at home or on the go when I may not have internet 24/7.  The configuration is also quite static.  I will only be using it on a specific machine with a specific setup, sure there are lots of permutations but VSCode is already doing the work of figuring out which files to download, it sounds like a simple change to cache that and used the cached copies if no updates are needed.  Additionally if it couldn't connect to the internet it would skip updating the files and just run as is.

**Comment by WSLUser** [comment]
> getting this working at home or on the go 

There are corporate and government networks as well that may be more lenient with the client OS but not the remote server. I have such as case for my use-case. Nobody developing at home should need to worry about it if it's personal network. That should simply be a matter of updating FW rules to allow traffic from the endpoints they specify in the docs.

**Comment by BMellor** [comment]
@strike5150: Don't be so sure about "most" users having connectivity at some point. Many corporations don't allow internet access except via a protected browser. Other applications are firewalled and can only access the internal network. Those that use VSCode at my work have to manually download update and extension files.

**Comment by yangxiang92** [comment]
Agree with @BMellor . A lot of corporations do not allow internet access because of information security concern. 
For me, I install all the required components on a computer with internet access and copy all the files related to vscode to another offline machine with the same enviroment as the previous one to make vscode-server possible to run on a offline machine. 
However, with current version of vscode-remote-ssh extention, it checks internet connection when the vscode-server is already installed. It makes my trick not working any more. I think this is not reasonable.

**Comment by roblourens** [comment]
> However, with current version of vscode-remote-ssh extention, it checks internet connection when the vscode-server is already installed

Can you open a new issue and explain some more, and include the log from the Remote-SSH output channel?

**Comment by yangxiang92** [comment]
I think i might have solved the problem. Thank you for your attention. @roblourens 

I reviewed the javascript code of the vscode-remote-ssh. It seems that the plugin is trying to use wget on the target machine to get something from the localhost, which is the target machine, to check if the server is working. But mine plugin seems to always get a invalid output from the wget command.

I double checked my vscode configuration on two machine (offline and online), and found that the offline machine has a vscode-remote-ssh plugin at version 0.45.2. However, the online machine, which is the one works fine, had a vscode-remote-ssh plugin at version 0.46.1. I upgraded the plugin of the offline machine to 0.46.1 and all problems were solved. The issue should be caused by the mis-match version of plugin and server, though the root-cause is not cleared yet.

**Comment by xiekeyi98** [comment]
I think we should have two settings:
1. Local host and remote host neither have internet.
2. Only remote host have not internet.( Because remote host is like IDC or Cloud ).

As 2 , Can we use local host as a proxy to install remote-ssh( And I think this can be default).
Indeed, I'am puzzled why we need remote host to access Internet? I think we can do everything by local host.

And another suggest:
I found it will use "SSH TCP FORWARD" , but in out company, "SSH FORWARD" without approval  is irregular, and we not use 22 as SSH port. Is there anyway to avoid ssh forward?

**Comment by roblourens** [comment]
2 is already supported, like the OP says. You can set `"remote.SSH.allowLocalServerDownload": true`.

**Comment by WSLUser** [comment]
A quick fix for now that should be enabled anyways: Fix this error: `Version mismatch, client refused.`

How is this error generated? Assuming someone manually grabbed the tar.gz manually, then installed it on the remote host: 

Update the extension,
Attempt to connect, it will fail.
Copy the contents of the previous server installed to the new commit id directory in `.vscode-server-insiders/bin`. Close ssh remote window and restart window. Watch it bypass the install as it detects an installation but fail to connect.

This error can been seen when looking at extensions. The actual output says something about curl failing. Seeing how the client refused to connect due to a hard-coded line for the version, this makes sense. The server should at least allow the last 5 versions to work before providing a prompt in a Windows Dialog preferably to install a new version with a link to the servers provided for quick easy one click download, which we could click on when disconnected from the corporate network. 

At that point we could then save it to wherever the server temporarily lives on client side, tell the extension to look for it when attempting to connect once the secure connection between client and host is in place. Discovering it, it will then proceed with the scp and installation process.

**Comment by IgnusG** [comment]
Is this the same as https://github.com/microsoft/vscode-remote-release/issues/1242?

https://github.com/microsoft/vscode-remote-release/issues/1802#issuecomment-555839078

**Comment by WSLUser** [comment]
I would consider it a sub task. Though I actually tested the recent Nightly and it appears that whenever you update VSCode, the issue I mentioned above (and also https://github.com/microsoft/vscode-remote-release/issues/1802#issuecomment-555839078 ) occurs. Updating the extension itself no longer seems to be the problem, perhaps it never was. So more accurately, VS Code should keep compatibility with the previous 5 servers. I think it's a check in the extension that verifies the server version is the same as what's reported by VSCode (Insiders) and so fails if they mismatch. IMO the only reason the server shouldn't work is because it becomes too incompatible with the VSCode client. Connecting itself should never be the problem.

**Comment by Coderx7** [comment]
Is there any update in this regard? should we be expecting something in near future(even possibly the next release) to address this ?

**Comment by fakhamatia** [comment]
I use [this](https://stackoverflow.com/a/56781109/10695748) trick for both side offline connection.
But when VSCode release an update commit id its change and I must download files again and upload on servers and extract tar.
I use VPNs to connect servers and when use VPN my internet disconnect, VSCode cant download new files.
Luckily extension install and update on servers without internet and complete offline connection.

**Comment by PavelSosin-320** [comment]
@fakhamatia Some WiFi routers running OpenWRT Unix has SSH Server package installed. Check your model in OpenWRT portal

**Comment by Spriithy** [comment]
Have there been any advancements towards using Remote SSH when neither hosts are connected to internet ? Thanks

**Comment by JP01** [comment]
Have there been any updates on this? Is it on the roadmap?
Thanks

**Comment by cleep55** [comment]
Where can I find older versions? I have an older version of vscode 1.52.1 on an offline network. I can't update vscode but would like to add the extension... I think it is 0.64.0_1 that I need...

**Comment by awesomebytes** [comment]
Just wanted to support this issue. I have been bitten by it multiple times already. I want to thank anyone that may be looking into it.

**Comment by skaravos** [comment]
+1
It has been over 3 years since this issue was opened, has there been any progress at all?

**Comment by aleksanm** [comment]
> +1 It has been over 3 years since this issue was opened, has there been any progress at all?

One way to get it done:
https://linuxtut.com/en/af61a2593f4ed1d4085c/

**Comment by max-sixty** [comment]
If anyone needs to do this — here's an approach that worked for me for the past couple of years:

- Download the server in an online system with `wget -q https://update.code.visualstudio.com/${VSCODE_VERSION}/server-linux-x64/stable -O - | tar -xz`; the version  should be like `1.67.1`. The versions are not compatible.
- Put the extension in the correct place in the offline system. Here's a Taskfile definition for doing this, but easy to turn into a bash script:
  ```yaml
  link-vscode-remote:
    desc: Copy the vscode remote extension. Run on the remote machine.
    # This is based on:
    # https://stackoverflow.com/a/57601121/3064736
    # https://stackoverflow.com/a/56781109/3064736
    vars:
      prefix: ~/.vscode-server/bin
      commit:
        sh: code --version | head -2 | tail -1
    cmds:
      - mkdir -p {{.prefix}}
      # Replace the dest with a link to our dir.
      - ln -sinf $PWD/vscode/vscode-server-linux-x64 {{.prefix}}/{{.commit}}
      # To avoid having to click on "Install on server" for each extenstion, we
      # link the server extensions to the client extensions.
      - ln -sinf $PWD/vscode/data/extensions ~/.vscode-server/extensions
  ```
  - (The extension linking in the final command requires a shared filesystem, but isn't required if you're OK doing lots of clicking)

**Comment by nasso** [comment]
upping again: this is a major inconvenience as I use dev containers to make programming on Windows acceptable. right now, even though the image is available locally, i am unable to start the dev container when im not connected to the internet.

**Comment by KholdStare** [comment]
I have already posted this comment in issue https://github.com/microsoft/vscode-remote-release/issues/1802, but duplicating here.

It is quite ridiculous that even though I have connected to the same Docker container hundreds of times before and should presumably have a cached copy of the VSCode server locally, I cannot connect to it when offline. The fact that I need an _internet connection_ for something that is _entirely local_ is mind boggling.

If I am on the go, in an airplane, or in a cafe and don't want to use the public Wifi, tools should work without a constant internet connection.

Please please fix this, this is **embarassing**!

**Comment by Svalkash** [comment]
> > +1 It has been over 3 years since this issue was opened, has there been any progress at all?
> 
> One way to get it done: https://linuxtut.com/en/af61a2593f4ed1d4085c/

It's still possible via modifying the JS file, but the link is outdated. You have to edit other lines right now.

**Comment by Crefok** [comment]
> > > +1 It has been over 3 years since this issue was opened, has there been any progress at all?
> > 
> > 
> > One way to get it done: https://linuxtut.com/en/af61a2593f4ed1d4085c/
> 
> It's still possible via modifying the JS file, but the link is outdated. You have to edit other lines right now.

The Link is working for me, but the described method isn't. I got stuck at the "Download VS Code Server..." Message...

**Comment by joshspicer** [comment]
For those asking about the status in the Remote - SSH extension: The `useLocalServer` setting should help in a majority of cases.  If the remote host isn't able to fetch the required server over the internet, it will attempt to `scp` the required server bits over from your local machine to the remote host. There seems to be some edge cases (https://github.com/microsoft/vscode-remote-release/issues/10344) and we'll look into those

I'm seeing some comments in this thread regarding dev containers - please redirect that feedback to https://github.com/microsoft/vscode-remote-release/issues/10173 so its seen by the relevant folks
